sdk/search/Azure.Search.Documents/src/Batching/InitializationBarrierSlim.cs

```diff
@@ -0,0 +1,92 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Azure.Search.Documents.Batching
+{
+    /// <summary>
+    /// InitializationBarrierSlim allows for synchronized lazy initialization.  An
+    /// AsyncLazy wouldn't work in these circumstances because we need the
+    /// results of the initialization before anyone else can proceed.
```

Comment:
Isn't this achieved by everyone awaiting the `await asyncLazy.Value`?

Reply:
Yes, I think we could probably make that work.